src/java/org/apache/cassandra/nodes/virtual/LegacySystemKeyspaceToNodes.java

```diff
@@ -0,0 +1,286 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.nodes.virtual;
+
+import java.io.IOException;
+import java.net.InetAddress;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.config.DatabaseDescriptor;
+import org.apache.cassandra.cql3.UntypedResultSet;
+import org.apache.cassandra.db.ColumnFamilyStore;
+import org.apache.cassandra.db.Keyspace;
+import org.apache.cassandra.db.SystemKeyspace;
+import org.apache.cassandra.db.commitlog.CommitLogPosition;
+import org.apache.cassandra.db.marshal.BytesType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.db.marshal.UUIDType;
+import org.apache.cassandra.dht.Token;
+import org.apache.cassandra.io.util.DataInputBuffer;
+import org.apache.cassandra.io.util.RebufferingInputStream;
+import org.apache.cassandra.locator.InetAddressAndPort;
+import org.apache.cassandra.nodes.BootstrapState;
+import org.apache.cassandra.nodes.LocalInfo;
+import org.apache.cassandra.nodes.NodeInfo;
+import org.apache.cassandra.nodes.Nodes;
+import org.apache.cassandra.schema.Schema;
+import org.apache.cassandra.schema.SchemaConstants;
+import org.apache.cassandra.service.StorageService;
+import org.apache.cassandra.utils.CassandraVersion;
+
+import static java.lang.String.format;
+import static org.apache.cassandra.cql3.QueryProcessor.executeInternal;
+import static org.apache.cassandra.nodes.virtual.NodeConstants.LEGACY_PEERS;
+import static org.apache.cassandra.nodes.virtual.NodeConstants.LOCAL;
+
+/**
+ * Upgrade code to migrate from {@code system.local} and {@code system.peers} to {@link Nodes}.
```

Comment:
The motivation is explained in https://datastax.jira.com/browse/DB-2394 but I have to admit I haven't gone thoroughly through it.
Such a design was introduced in DSE 6.8 and it seems it remained in CC as well - likely because Astra had to migrate from 6.8 too. 
Given that HCD relies on CC quite heavily I didn't consider reverting it, doing so would require a major effort and would complicate the upgrade from 6.8 as well.

Reply:
I wouldn't be thinking of reverting it, so much as just not enabling that behaviour (until it is actually needed).  But that's the cart before the horse – right now why do we need this ?

Reply:
I can have a look at how to make CC read the data from raw sstables instead but I have the feeling it is going only to delay the problem not solve it. It is going to have another consequence: upgrading from 5.1 your system data in HCD will be stored in sstables while when upgrading from 6.8 it is going to be in flat files - a bit messy state to end up with.

Reply:
HCD is supposed to be converged to Cassandra, or at least close.  Why not just read from raw sstables if there's, in the product, no user value provided from the flat files? 

These flat files, sound to me like, a DSE/Astra feature. What the actual value now, and coming ?

The upgrade from 6.8 is a valid point, but also what a a user going from HCD to C* ?  And what about the new code, and new matrix of upgrade code when upgrading to a TCM based version…

Getting a statement of these questions and our current answers is IMHO more important than what lands in this PR. (for reference sake)

Reply:
I will have to reach out to people with better CC backgrounds to find these answers.
Do we need to consider upgrading HCD to C*?
Regarding TCM, I know nothing about it, how does it affect system tables?

Reply:
> Do we need to consider upgrading HCD to C*?

IMHO, this deserves separate engineering and product answers.  Product-wise I doubt anything beyond an offline upgrade has much value. 

> Regarding TCM, I know nothing about it, how does it affect system tables? 

For the most part builds its constructs on top… 
see 
- https://github.com/apache/cassandra/commit/ae0842372ff6dd1437d026f82968a3749f555ff4#diff-f57518f964c71328146aeca95be5e697ca81a77261719eeef4dd4b1ed8daf63b
- https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/tcm/ClusterMetadata.java

Reply:
We need the node construct in Astra.

Reply:
> We need the node construct in Astra.

Can you elaborate please @JeremiahDJordan .

What's it needed for ? Why wasn't it there already ?

Reply:
It is already there.  You sound like you are wanting to remove it.  This PR is not adding the Node concept, this PR is adding the ability for it to read from legacy system tables which is required for DSE 5.1 compatibility.

Reply:
Ah yes, you're right.  All that's been added here is the upgrade to it.

Why does Astra need it?  Is this needed by, or of value to, HCD ?

Reply:
Generally, i'm aligned with Josh's earlier comment on that ticket, here:
https://datastax.jira.com/browse/DB-2394?focusedCommentId=311781 

if HCD is going to be the sole owners of this, and the TPC related benefits are not longer present, and it's going to be something different anyway in TCM, let's lighten our load and surface area for bugs.  (and yes, it's a shame it's in HCD 1.0 already)